src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs

```diff
@@ -182,6 +186,95 @@ public static List<string> GetHeaderValues(HttpRequestHeaders headers, string he
+            Func<HttpResponseMessage, bool> validationCallback = null, int retryCount = 5, int retryDelaySeconds = 5)
+        {
+            if (loggingHelper == null)
+                throw new ArgumentNullException(nameof(loggingHelper));
+            if (send == null)
+                throw new ArgumentNullException(nameof(send));
+            if (retryCount < 1)
+                throw new ArgumentException(nameof(retryCount));
+            if (retryDelaySeconds < 1)
+                throw new ArgumentException(nameof(retryDelaySeconds));
+
+            int retries = 0;
+            HttpResponseMessage response = null;
+
+            // add a bit of randomness to the retry delay
+            var rng = new Random();
+
+            while (retries < retryCount)
+            {
+                if (retries > 1)
+                {
+                    int delay = retryDelaySeconds * (retries - 1) * rng.Next(1, 5);
+                    loggingHelper.LogMessage("Waiting {0} seconds before retry", delay);
+                    Thread.Sleep(delay * 1000);
+                }
+
+                try
+                {
+                    response = await send();
+                }
+                catch (Exception e)
+                {
+                    loggingHelper.LogWarningFromException(e, true);
+
+                    // if this is the final iteration let the exception bubble up
+                    if (retries + 1 == retryCount)
+                        throw;
+                }
+
+                if (validationCallback == null)
+                {
+                    // check if the response code is within the range of failures
+                    if (IsWithinRetryRange(response.StatusCode))
+                    {
+                        loggingHelper.LogWarning("Request failed with status code {0}", response.StatusCode);
+                    }
+                    else
+                    {
+                        loggingHelper.LogMessage(MessageImportance.Low, "Response completed with status code {0}", response.StatusCode);
+                        return response;
+                    }
+                }
+                else
+                {
+                    bool isSuccess = validationCallback(response);
+                    if (!isSuccess)
+                    {
+                        loggingHelper.LogMessage("Validation callback returned retry for status code {0}", response.StatusCode);
+                    }
+                    else
+                    {
+                        loggingHelper.LogMessage("Validation callback returned success for status code {0}", response.StatusCode);
+                        return response;
+                    }
+                }
+
+                ++retries;
+            }
+
+            // retry count exceeded
+            loggingHelper.LogWarning("Retry count {0} exceeded", retryCount);
+            throw new Exception(string.Format("Request failed with status {0} response {1}",
```

Comment:
I wish we could throw something more useful.  Forcing people to do "catch(Exception)" isn't ideal.